File: python/train.py
#!/usr/bin/env python3
"""
Drakon AI - Python Training Interface

This module provides a high-level Python interface for training neural networks
using the Drakon AI C++ backend.
"""

import argparse
import json
import os
import sys
import time
from typing import Dict, Any, Optional, Tuple
import numpy as np

# Try to import the C++ bindings
try:
    import drakon_python as drakon
except ImportError:
    print("Warning: C++ bindings not available. Using mock implementation.")
    drakon = None

class DrakonTrainer:
    """High-level trainer class for Drakon AI models."""
    
    def __init__(self, config: Dict[str, Any]):
        """Initialize the trainer with configuration."""
        self.config = config
        self.model = None
        self.optimizer = None
        self.train_loader = None
        self.val_loader = None
        
    def create_model(self) -> None:
        """Create the neural network model."""
        if drakon:
            # Use C++ backend
            layer_sizes = self.config['network']['layer_sizes']
            activation = self.config['network']['activation_type']
            self.model = drakon.MLP(layer_sizes, activation)
        else:
            # Mock implementation
            print("Creating mock model...")
            self.model = MockModel(self.config)
    
    def create_optimizer(self) -> None:
        """Create the optimizer."""
        if drakon:
            opt_type = self.config['training']['optimizer_type']
            lr = self.config['training']['learning_rate']
            self.optimizer = drakon.create_optimizer(opt_type, lr)
        else:
            print("Creating mock optimizer...")
            self.optimizer = MockOptimizer(self.config)
    
    def create_data_loaders(self) -> None:
        """Create training and validation data loaders."""
        if drakon:
            # Use C++ backend
            dataset_config = self.config['dataset']
            train_dataset = drakon.SyntheticDataset(
                dataset_config['num_samples'],
                dataset_config['input_features'],
                dataset_config['num_classes'],
                dataset_config['problem_type'],
                True
            )
            self.train_loader = drakon.DataLoader(
                train_dataset,
                self.config['training']['batch_size'],
                True
            )
        else:
            # Mock implementation
            print("Creating mock data loaders...")
            self.train_loader = MockDataLoader(self.config)
            self.val_loader = MockDataLoader(self.config)
    
    def train_epoch(self) -> Tuple[float, float]:
        """Train for one epoch."""
        if not self.model or not self.optimizer or not self.train_loader:
            raise RuntimeError("Model, optimizer, or data loader not initialized")
        
        total_loss = 0.0
        total_accuracy = 0.0
        num_batches = 0
        
        for batch_idx in range(self.train_loader.num_batches()):
            batch_data, batch_labels = self.train_loader.get_batch(batch_idx)
            
            # Forward pass
            predictions = self.model.forward(batch_data)
            
            # Compute loss
            if self.config['dataset']['problem_type'] == 'classification':
                loss = drakon.cross_entropy_loss(predictions, batch_labels) if drakon else 0.1
                accuracy = drakon.accuracy(predictions, batch_labels) if drakon else 0.8
            else:
                loss = drakon.mse_loss(predictions, batch_labels) if drakon else 0.1
                accuracy = 0.0
            
            total_loss += loss
            total_accuracy += accuracy
            num_batches += 1
            
            # Backward pass
            grad_output = predictions - batch_labels  # Simplified gradient
            self.model.backward(grad_output)
            
            # Update parameters
            self.optimizer.step()
            self.optimizer.zero_grad()
        
        avg_loss = total_loss / num_batches
        avg_accuracy = total_accuracy / num_batches
        
        return avg_loss, avg_accuracy
    
    def validate(self) -> Tuple[float, float]:
        """Validate the model."""
        if not self.model or not self.val_loader:
            return 0.0, 0.0
        
        total_loss = 0.0
        total_accuracy = 0.0
        num_batches = 0
        
        for batch_idx in range(self.val_loader.num_batches()):
            batch_data, batch_labels = self.val_loader.get_batch(batch_idx)
            
            # Forward pass only
            predictions = self.model.forward(batch_data)
            
            # Compute metrics
            if self.config['dataset']['problem_type'] == 'classification':
                loss = drakon.cross_entropy_loss(predictions, batch_labels) if drakon else 0.1
                accuracy = drakon.accuracy(predictions, batch_labels) if drakon else 0.8
            else:
                loss = drakon.mse_loss(predictions, batch_labels) if drakon else 0.1
                accuracy = 0.0
            
            total_loss += loss
            total_accuracy += accuracy
            num_batches += 1
        
        avg_loss = total_loss / num_batches
        avg_accuracy = total_accuracy / num_batches
        
        return avg_loss, avg_accuracy
    
    def train(self) -> Dict[str, Any]:
        """Main training loop."""
        print("=== Starting Training ===")
        
        # Initialize components
        self.create_model()
        self.create_optimizer()
        self.create_data_loaders()
        
        # Training loop
        epochs = self.config['training']['epochs']
        best_loss = float('inf')
        patience_counter = 0
        patience = self.config['training']['early_stopping_patience']
        
        history = {
            'train_loss': [],
            'train_accuracy': [],
            'val_loss': [],
            'val_accuracy': []
        }
        
        start_time = time.time()
        
        for epoch in range(epochs):
            epoch_start = time.time()
            
            # Training
            train_loss, train_accuracy = self.train_epoch()
            
            # Validation
            val_loss, val_accuracy = self.validate()
            
            epoch_time = time.time() - epoch_start
            
            # Store history
            history['train_loss'].append(train_loss)
            history['train_accuracy'].append(train_accuracy)
            history['val_loss'].append(val_loss)
            history['val_accuracy'].append(val_accuracy)
            
            # Print progress
            if epoch % 10 == 0 or epoch == epochs - 1:
                print(f"Epoch {epoch+1}/{epochs} - "
                      f"Train Loss: {train_loss:.4f}, Train Acc: {train_accuracy:.4f}, "
                      f"Val Loss: {val_loss:.4f}, Val Acc: {val_accuracy:.4f}, "
                      f"Time: {epoch_time:.2f}s")
            
            # Early stopping
            if val_loss < best_loss - self.config['training']['min_delta']:
                best_loss = val_loss
                patience_counter = 0
                
                # Save best model
                if self.config['model']['save_best_only']:
                    self.save_model(f"best_model_{epoch+1}.bin")
            else:
                patience_counter += 1
                if patience_counter >= patience:
                    print(f"Early stopping triggered after {epoch + 1} epochs")
                    break
        
        total_time = time.time() - start_time
        
        # Final results
        results = {
            'final_train_loss': history['train_loss'][-1],
            'final_train_accuracy': history['train_accuracy'][-1],
            'final_val_loss': history['val_loss'][-1],
            'final_val_accuracy': history['val_accuracy'][-1],
            'best_val_loss': best_loss,
            'total_training_time': total_time,
            'epochs_trained': len(history['train_loss']),
            'history': history
        }
        
        print(f"\n=== Training Complete ===")
        print(f"Total time: {total_time:.2f} seconds")
        print(f"Best validation loss: {best_loss:.4f}")
        print(f"Final validation accuracy: {results['final_val_accuracy']:.4f}")
        
        return results
    
    def save_model(self, filename: str) -> None:
        """Save the model to file."""
        if self.model and drakon:
            drakon.save_model(self.model, filename)
        else:
            print(f"Mock: Saving model to {filename}")
    
    def load_model(self, filename: str) -> None:
        """Load the model from file."""
        if drakon:
            self.model = drakon.load_model(filename)
        else:
            print(f"Mock: Loading model from {filename}")


class MockModel:
    """Mock model for testing when C++ bindings are not available."""
    
    def __init__(self, config: Dict[str, Any]):
        self.config = config
        self.num_parameters = sum(config['network']['layer_sizes'][i] * config['network']['layer_sizes'][i+1] 
                                 for i in range(len(config['network']['layer_sizes'])-1))
    
    def forward(self, x):
        # Mock forward pass
        batch_size = x.shape[0] if hasattr(x, 'shape') else 1
        output_size = self.config['network']['layer_sizes'][-1]
        return np.random.rand(batch_size, output_size)
    
    def backward(self, grad_output):
        # Mock backward pass
        return np.random.rand(*grad_output.shape)


class MockOptimizer:
    """Mock optimizer for testing when C++ bindings are not available."""
    
    def __init__(self, config: Dict[str, Any]):
        self.config = config
        self.learning_rate = config['training']['learning_rate']
    
    def step(self):
        # Mock parameter update
        pass
    
    def zero_grad(self):
        # Mock gradient zeroing
        pass


class MockDataLoader:
    """Mock data loader for testing when C++ bindings are not available."""
    
    def __init__(self, config: Dict[str, Any]):
        self.config = config
        self.batch_size = config['training']['batch_size']
        self.num_samples = config['dataset']['num_samples']
        self.input_features = config['dataset']['input_features']
        self.num_classes = config['dataset']['num_classes']
    
    def num_batches(self) -> int:
        return (self.num_samples + self.batch_size - 1) // self.batch_size
    
    def get_batch(self, batch_idx: int) -> Tuple[np.ndarray, np.ndarray]:
        # Generate mock data
        batch_size = min(self.batch_size, self.num_samples - batch_idx * self.batch_size)
        data = np.random.rand(batch_size, self.input_features)
        
        if self.config['dataset']['problem_type'] == 'classification':
            labels = np.zeros((batch_size, self.num_classes))
            for i in range(batch_size):
                labels[i, np.random.randint(0, self.num_classes)] = 1.0
        else:
            labels = np.random.rand(batch_size, 1)
        
        return data, labels


def load_config(config_file: str) -> Dict[str, Any]:
    """Load configuration from file."""
    if not os.path.exists(config_file):
        # Create default configuration
        config = {
            'network': {
                'layer_sizes': [10, 64, 32, 2],
                'activation_type': 'relu',
                'output_activation': 'softmax',
                'use_dropout': True,
                'dropout_rate': 0.3,
                'weight_init': 'xavier'
            },
            'training': {
                'epochs': 100,
                'batch_size': 32,
                'learning_rate': 0.001,
                'optimizer_type': 'adam',
                'momentum': 0.9,
                'beta1': 0.9,
                'beta2': 0.999,
                'alpha': 0.99,
                'epsilon': 1e-8,
                'shuffle_data': True,
                'validation_split': 0.2,
                'early_stopping_patience': 10,
                'min_delta': 1e-4
            },
            'dataset': {
                'type': 'synthetic',
                'num_samples': 1000,
                'input_features': 10,
                'num_classes': 2,
                'problem_type': 'classification',
                'file_path': '',
                'has_header': True,
                'normalize_data': False,
                'standardize_data': True
            },
            'model': {
                'save_path': 'model.bin',
                'load_path': '',
                'save_best_only': True,
                'save_format': 'binary',
                'save_optimizer_state': True
            }
        }
        
        # Save default configuration
        with open(config_file, 'w') as f:
            json.dump(config, f, indent=2)
        
        print(f"Created default configuration: {config_file}")
    else:
        # Load existing configuration
        with open(config_file, 'r') as f:
            config = json.load(f)
    
    return config


def main():
    """Main training script."""
    parser = argparse.ArgumentParser(description='Drakon AI Training Script')
    parser.add_argument('--config', type=str, default='config.json',
                       help='Configuration file path')
    parser.add_argument('--model-path', type=str, default='',
                       help='Path to load pre-trained model')
    parser.add_argument('--output-dir', type=str, default='./output',
                       help='Output directory for results')
    
    args = parser.parse_args()
    
    # Load configuration
    config = load_config(args.config)
    
    # Create output directory
    os.makedirs(args.output_dir, exist_ok=True)
    
    # Create trainer
    trainer = DrakonTrainer(config)
    
    # Load pre-trained model if specified
    if args.model_path:
        trainer.load_model(args.model_path)
    
    # Train the model
    results = trainer.train()
    
    # Save results
    results_file = os.path.join(args.output_dir, 'training_results.json')
    with open(results_file, 'w') as f:
        json.dump(results, f, indent=2)
    
    print(f"Results saved to: {results_file}")
    
    # Save final model
    final_model_path = os.path.join(args.output_dir, 'final_model.bin')
    trainer.save_model(final_model_path)
    print(f"Final model saved to: {final_model_path}")


if __name__ == '__main__':
    main()

File: src/train.cpp
#include "config.h"
#include "dataset.h"
#include "layers.h"
#include "optimizer.h"
#include "utils.h"
#include <iostream>
#include <memory>
#include <string>

int main(int argc, char** argv) {
    try {
        // Parse command line arguments
        std::string config_file = "config.txt";
        if (argc > 1) {
            config_file = argv[1];
        }
        
        std::cout << "=== Drakon AI Training ===" << std::endl;
        std::cout << "Loading configuration from: " << config_file << std::endl;
        
        // Load configuration
        Config config;
        if (!config.load_from_file(config_file)) {
            std::cout << "Creating default configuration..." << std::endl;
            config = create_default_config("classification", 10, 2);
            config.save_to_file(config_file);
        }
        
        // Validate configuration
        if (!config.validate()) {
            std::cerr << "Invalid configuration!" << std::endl;
            return 1;
        }
        
        config.print_summary();
        
        // Create dataset
        std::cout << "\n=== Creating Dataset ===" << std::endl;
        auto dataset = std::make_shared<SyntheticDataset>(
            config.get_dataset_config().num_samples,
            config.get_dataset_config().input_features,
            config.get_dataset_config().num_classes,
            config.get_dataset_config().problem_type,
            config.get_training_config().shuffle_data
        );
        
        std::cout << "Dataset created: " << dataset->size() << " samples" << std::endl;
        std::cout << "Input dimension: " << dataset->input_dim() << std::endl;
        std::cout << "Output dimension: " << dataset->output_dim() << std::endl;
        
        // Create model
        std::cout << "\n=== Creating Model ===" << std::endl;
        auto model = std::make_unique<MLP>(
            config.get_network_config().layer_sizes,
            config.get_network_config().activation_type,
            "main_model"
        );
        
        std::cout << "Model created with " << model->num_parameters() << " parameters" << std::endl;
        
        // Create optimizer
        std::cout << "\n=== Creating Optimizer ===" << std::endl;
        auto optimizer = create_optimizer(
            config.get_training_config().optimizer_type,
            config.get_training_config().learning_rate
        );
        
        // Set optimizer parameters
        auto params = model->get_parameters();
        auto grads = model->get_gradients();
        optimizer->set_parameters(params, grads);
        
        std::cout << "Optimizer created: " << config.get_training_config().optimizer_type << std::endl;
        
        // Create data loader
        DataLoader train_loader(dataset, config.get_training_config().batch_size, true);
        
        // Training loop
        std::cout << "\n=== Starting Training ===" << std::endl;
        utils::timing::Timer timer;
        timer.start();
        
        float best_loss = std::numeric_limits<float>::infinity();
        size_t patience_counter = 0;
        
        for (size_t epoch = 0; epoch < config.get_training_config().epochs; ++epoch) {
            float epoch_loss = 0.0f;
            size_t num_batches = 0;
            
            // Training epoch
            for (size_t batch_idx = 0; batch_idx < train_loader.num_batches(); ++batch_idx) {
                auto [batch_data, batch_labels] = train_loader.get_batch(batch_idx);
                
                // Forward pass
                Tensor predictions = model->forward(batch_data);
                
                // Compute loss
                float loss;
                if (config.get_dataset_config().problem_type == "classification") {
                    loss = utils::loss::cross_entropy_loss(predictions, batch_labels);
                } else {
                    loss = utils::loss::mse_loss(predictions, batch_labels);
                }
                
                epoch_loss += loss;
                num_batches++;
                
                // Backward pass
                Tensor grad_output = predictions - batch_labels;  // Simplified gradient
                model->backward(grad_output);
                
                // Update parameters
                optimizer->step();
                optimizer->zero_grad();
            }
            
            float avg_loss = epoch_loss / num_batches;
            
            // Print progress
            if (epoch % 10 == 0 || epoch == config.get_training_config().epochs - 1) {
                std::cout << "Epoch " << epoch << "/" << config.get_training_config().epochs 
                         << " - Loss: " << avg_loss << std::endl;
            }
            
            // Early stopping
            if (avg_loss < best_loss - config.get_training_config().min_delta) {
                best_loss = avg_loss;
                patience_counter = 0;
                
                // Save best model
                if (config.get_model_config().save_best_only) {
                    // In a real implementation, you'd save the model here
                    std::cout << "New best loss: " << best_loss << " (saving model)" << std::endl;
                }
            } else {
                patience_counter++;
                if (patience_counter >= config.get_training_config().early_stopping_patience) {
                    std::cout << "Early stopping triggered after " << epoch + 1 << " epochs" << std::endl;
                    break;
                }
            }
        }
        
        timer.stop();
        std::cout << "\n=== Training Complete ===" << std::endl;
        std::cout << "Total training time: " << timer.elapsed_seconds() << " seconds" << std::endl;
        std::cout << "Best loss achieved: " << best_loss << std::endl;
        
        // Test the model
        std::cout << "\n=== Testing Model ===" << std::endl;
        dataset->reset();
        auto [test_data, test_labels] = dataset->get_batch(100);
        Tensor test_predictions = model->forward(test_data);
        
        float test_loss;
        if (config.get_dataset_config().problem_type == "classification") {
            test_loss = utils::loss::cross_entropy_loss(test_predictions, test_labels);
            float accuracy = utils::metrics::accuracy(test_predictions, test_labels);
            std::cout << "Test Loss: " << test_loss << std::endl;
            std::cout << "Test Accuracy: " << accuracy << std::endl;
        } else {
            test_loss = utils::loss::mse_loss(test_predictions, test_labels);
            float r2 = utils::metrics::r_squared(test_predictions, test_labels);
            std::cout << "Test Loss: " << test_loss << std::endl;
            std::cout << "Test R²: " << r2 << std::endl;
        }
        
        std::cout << "\n=== Training Summary ===" << std::endl;
        std::cout << "✅ Training completed successfully!" << std::endl;
        std::cout << "✅ Model is ready for inference" << std::endl;
        
        return 0;
        
    } catch (const std::exception& e) {
        std::cerr << "Error during training: " << e.what() << std::endl;
        return 1;
    }
}
